src/datatablerow/datatablerow.jl

```diff
@@ -37,31 +37,115 @@ Base.done(r::DataTableRow, s) = s > length(r)
+            eq_col = convert(Nullable{Bool}, col1[r1.row] == col2[r2.row])
+            # If true or null, need to compare remaining columns
+            get(eq_col, true) || return Nullable(false)
+            eq &= eq_col
+        end
+        return eq
+    else
+        r1.row == r2.row && return Nullable(true)
+        eq = Nullable(true)
+        @inbounds for col in columns(r1.dt)
+            eq_col = convert(Nullable{Bool}, col[r1.row] == col[r2.row])
+            # If true or null, need to compare remaining columns
+            get(eq_col, true) || return Nullable(false)
+            eq &= eq_col
         end
+        return eq
+    end
+end
+
+# internal method for comparing the elements of the same data table column
+isequal_colel(col::AbstractArray, r1::Int, r2::Int) =
+    (r1 == r2) || isequal(Base.unsafe_getindex(col, r1), Base.unsafe_getindex(col, r2))
+
+function isequal_colel{T}(col::Union{NullableArray{T},
+                                     AbstractNullableCategoricalArray{T}},
+                                     r1::Int, r2::Int)
+    (r1 == r2) && return true
+    isequal(col[r1], col[r2])
+end
+
+isequal_colel(a::Any, b::Any) = isequal(a, b)
+isequal_colel(a::Nullable, b::Any) = !isnull(a) & isequal(unsafe_get(a), b)
+isequal_colel(a::Any, b::Nullable) = isequal_colel(b, a)
+isequal_colel(a::Nullable, b::Nullable) = isequal(a, b)
+
+# comparison of DataTable rows
+function isequal_row(dt::AbstractDataTable, r1::Int, r2::Int)
+    (r1 == r2) && return true # same row
+    @inbounds for col in columns(dt)
+        isequal_colel(col, r1, r2) || return false
+    end
+    return true
+end
+
+function isequal_row(dt1::AbstractDataTable, r1::Int, dt2::AbstractDataTable, r2::Int)
+    (dt1 === dt2) && return isequal_row(dt1, r1, r2)
+    (ncol(dt1) == ncol(dt2)) ||
+        throw(ArgumentError("Rows of the data tables that have different number of columns cannot be compared ($(ncol(dt1)) and $(ncol(dt2)))"))
+    @inbounds for (col1, col2) in zip(columns(dt1), columns(dt2))
+        isequal_colel(col1[r1], col2[r2]) || return false
     end
     return true
 end
+
+# comparison of DataTable rows
+# rows are equal if they have the same values (while the row indices could differ)
+Base.isequal(r1::DataTableRow, r2::DataTableRow) =
+    isequal_row(r1.dt, r1.row, r2.dt, r2.row)
+
+# lexicographic ordering on DataTable rows, null > !null
+function Base.isless(r1::DataTableRow, r2::DataTableRow)
+    (ncol(r1.dt) == ncol(r2.dt)) ||
+        throw(ArgumentError("Rows of the data tables that have different number of columns cannot be compared ($(ncol(dt1)) and $(ncol(dt2)))"))
+    @inbounds for i in 1:ncol(r1.dt)
+        col1 = r1.dt[i]
+        col2 = r2.dt[i]
+        isnull1 = _isnull(col1[r1.row])
+        isnull2 = _isnull(col2[r2.row])
+        (isnull1 != isnull2) && return isnull2 # null > !null
+        if !isnull1
+            v1 = get(col1[r1.row])
```

Comment:
I think you'll need a `_get` function, just like `_isnull`, which will be a no-op for non nullables. Looks like we also need a test with non-nullable columns to catch this (unless I'm missing something).